src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+implemented as functions, which server as their ``render()`` method).]
+
+React can be used without JSX, but it saves a lot of verbosity and is
+probably a bit clearer, especially to anyone with HTML template experience.
+
+CJSX
+^^^^
+
+If JSX is the preferred way to write React components, this presents a
+challenge for integrating JSX with codebases that otherwise use
+CoffeeScript.  One could write everything in CoffeeScript *except* for the
+code for React components (which would have to go in separate JSX files),
+but that introduces another difficult cognitive overhead.
+
+To solve that, the `CJSX`_ language is just a simple superset of CoffeeScript
+to support JSX-like syntax.  In other words, CJSX is to CoffeeScript as JSX is
+to vanilla JavaScript.
+
+So this is what all the ``.cjsx`` sources (something that might be new to
+most readers) are in SMC.  If you see a file in SMC with the ``.cjsx``
+extension you can bet there's probably a React component defined in there
+somewhere.
+
+.. note:
+    In the process of researching this we learned that the original
+    developer of CJSX has abandoned the project and there isn't really
+    anything to take its place yet.  William `insists
+    <https://github.com/sagemathinc/smc/issues/1545>`_ that SMC will
+    continue to use it, and with good reason!, but it leaves me not without
+    doubts...
+
+Redux
+-----
+
+It's a little tricky to explain exactly what Redux 'is' without specific
+examples.  According to its `docs <http://redux.js.org/>`_:
+
+    Redux is a predictable state container for JavaScript apps.
+
+It's really little more than a simple protocol for application state
+updates by way of immutable state containers and pure functions that return
+an updated state based on some action performed on it (where the action
+can be any abstract operation that results in an updated application
+state). These functions are called reducers.
+
+There's very little else in Redux--it's mostly convenience functions for
+managing a state object, and combining reducers to produce new states from
+state changing actions.
+
+The purpose, all in all, is to provide a sane, predictable, reproducible
+way to manage and track (using middleware that logs actions) the live
+state of a complex application.  We'll come back to this later with some
+specific examples.  SMC wraps most of its use of Redux into its own
+abstractions that are implemented in ``smc-webapp/smc-react.coffee``.
+
+React-Redux
+^^^^^^^^^^^
+
+SMC's ``smc-react.coffee`` modules also makes use of the `React Redux`_
+JavaScript module to tie Redux state objects to React containers (i.e. update
+displays when the state changes--abstracting the state itself from any given
+view of the state).  This is just a package for making it convenient to
+implement model / view separation in React components.  The web developers
+describe this as "container components" but really they're just reinventing MVC
```

Comment:
I see what you're saying here.  I'm being glib when I write that it's "[just] reinventing MVC".  There are plenty of different places to draw the line, for example, between presentation and data, and I've been in several debates about that with no clear right answer.

I think my comments here just stem from frustration with some of the jargon I encounter when reading about web frameworks--especially ones like React.  So when I read in some docs somewhere something like "this is so that you can do such and such with your container components".  Then I try to google about what they mean by "container components" and get some long-winded explanation that could have been summarized (to me) as something like "we do things like this sometimes to keep purely presentational components independent of the application state" which seems obvious to me.

That said, I should go through and remove little subjective comments like these--like I said, these started as my personal notes :)